zebra-test/src/command.rs

```diff
@@ -248,20 +185,130 @@ impl TestOutput {
+    #[cfg(not(unix))]
+    fn was_killed(&self) -> bool {
+        self.output.status.code() != Some(1)
+    }
 
-        #[cfg(not(unix))]
-        return self.output.status.code() == Some(1);
+    #[cfg(unix)]
+    fn was_killed(&self) -> bool {
+        self.output.status.signal() != Some(9)
+    }
+}
+
+/// Add context to an error report
+trait ContextFrom<S> {
+    type Return;
+
+    fn context_from(self, source: &S) -> Self::Return;
+}
+
+impl<C, T, E> ContextFrom<C> for Result<T, E>
+where
+    E: Into<Report>,
+    Report: ContextFrom<C, Return = Report>,
+{
+    type Return = Result<T, Report>;
+
+    fn context_from(self, source: &C) -> Self::Return {
+        self.map_err(|e| e.into())
+            .map_err(|report| report.context_from(source))
+    }
+}
+
+impl ContextFrom<TestStatus> for Report {
+    type Return = Report;
+
+    fn context_from(self, source: &TestStatus) -> Self::Return {
+        let command = || source.cmd.clone().header("Command:");
+
+        self.with_section(command).context_from(&source.status)
+    }
+}
+
+impl ContextFrom<TestChild> for Report {
+    type Return = Report;
+
+    fn context_from(self, source: &TestChild) -> Self::Return {
+        let command = || source.cmd.clone().header("Command:");
+        let child = || format!("{:?}", source.child).header("Child Process:");
+
+        self.with_section(command).with_section(child)
+    }
+}
+
+impl ContextFrom<TestOutput> for Report {
+    type Return = Report;
+
+    fn context_from(self, source: &TestOutput) -> Self::Return {
+        self.with_section(|| source.cmd.clone().header("Command:"))
+            .context_from(&source.output)
+    }
+}
+
+impl ContextFrom<Output> for Report {
+    type Return = Report;
+
+    fn context_from(self, source: &Output) -> Self::Return {
+        let stdout = || {
+            String::from_utf8_lossy(&source.stdout)
+                .into_owned()
+                .header("Stdout:")
+        };
+        let stderr = || {
+            String::from_utf8_lossy(&source.stderr)
+                .into_owned()
+                .header("Stderr:")
+        };
+
+        self.with_section(stdout)
+            .with_section(stderr)
+            .context_from(&source.status)
+    }
+}
+
+impl ContextFrom<ExitStatus> for Report {
+    type Return = Report;
+
+    fn context_from(self, source: &ExitStatus) -> Self::Return {
+        let how = if source.success() {
+            "successfully"
+        } else {
+            "unsuccessfully"
+        };
+
+        if let Some(code) = source.code() {
+            self.with_section(|| {
+                format!("command exited {} with status code {}", how, code).header("Exit Status:")
+            })
+        } else {
+            self.with_section(|| {
+                format!("command exited {} without a status code", how).header("Exit Status:")
+            })
+            .note("processes exit without a status code on unix if terminated by a signal")
```

Comment:
Can we add the signal number on unix platforms?

It could be useful for diagnostics, if we ever get weird test failures due to out of memory errors, or other signals.

Something like:
```rust
#[cfg(unix)]
if let Some(signal) = self.output.status.signal() {
   format!("command exited {} with unix signal {}", how, signal).header("Exit Status:")
} else {
    format!("command exited {} without a status code or unix signal", how).header("Exit Status:")
}

#[cfg(not(unix))]
format!("command exited {} without a status code", how).header("Exit Status:")
```

Reply:
hell yeah we can